src/server/worker/utils/utils.go

```diff
@@ -0,0 +1,662 @@
+			result.uid = &uid32
+			gid, err := strconv.ParseUint(user.Gid, 10, 32)
+			if err != nil {
+				return nil, err
+			}
+			gid32 := uint32(gid)
+			result.gid = &gid32
+		}
+	}
+
+	return result, nil
+}
+
+// lookupDockerUser looks up users given the argument to a Dockerfile USER directive.
+// According to Docker's docs this directive looks like:
+// USER <user>[:<group>] or
+// USER <UID>[:<GID>]
+func lookupDockerUser(userArg string) (_ *user.User, retErr error) {
+	userParts := strings.Split(userArg, ":")
+	userOrUID := userParts[0]
+	groupOrGID := ""
+	if len(userParts) > 1 {
+		groupOrGID = userParts[1]
+	}
+	passwd, err := os.Open("/etc/passwd")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := passwd.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	scanner := bufio.NewScanner(passwd)
+	for scanner.Scan() {
+		parts := strings.Split(scanner.Text(), ":")
+		if parts[0] == userOrUID || parts[2] == userOrUID {
+			result := &user.User{
+				Username: parts[0],
+				Uid:      parts[2],
+				Gid:      parts[3],
+				Name:     parts[4],
+				HomeDir:  parts[5],
+			}
+			if groupOrGID != "" {
+				if parts[0] == userOrUID {
+					// groupOrGid is a group
+					group, err := lookupGroup(groupOrGID)
+					if err != nil {
+						return nil, err
+					}
+					result.Gid = group.Gid
+				} else {
+					// groupOrGid is a gid
+					result.Gid = groupOrGID
+				}
+			}
+			return result, nil
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Fatal(err)
+	}
+	return nil, fmt.Errorf("user %s not found", userArg)
+}
+
+func lookupGroup(group string) (_ *user.Group, retErr error) {
+	groupFile, err := os.Open("/etc/group")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := groupFile.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	scanner := bufio.NewScanner(groupFile)
+	for scanner.Scan() {
+		parts := strings.Split(scanner.Text(), ":")
+		if parts[0] == group {
+			return &user.Group{
+				Gid:  parts[2],
+				Name: parts[0],
+			}, nil
+		}
+	}
+	return nil, fmt.Errorf("group %s not found", group)
+}
+
+func (u *utils) Jobs() col.Collection {
+	return u.jobs
+}
+
+func (u *utils) Pipelines() col.Collection {
+	return u.pipelines
+}
+
+func (u *utils) Chunks(jobID string) col.Collection {
+	return col.NewCollection(u.etcdClient, path.Join(u.etcdPrefix, chunkPrefix, jobID), nil, &common.ChunkState{}, nil, nil)
+}
+
+func (u *utils) Merges(jobID string) col.Collection {
+	return col.NewCollection(u.etcdClient, path.Join(u.etcdPrefix, mergePrefix, jobID), nil, &common.MergeState{}, nil, nil)
+}
+
+// Create puller *
+// downloadData *
+// defer removeAll *
+// defer puller.Cleanup
+// lock runMutex
+// lock status mutex, set some dumb values
+// mkdir
+// linkData *
+// defer unlinkData
+// walk pfs and chown
+// run user code and user error handling code
+// puller.Cleanup
+// report stats
+// uploadOutput
+
+func (u *utils) WithProvisionedNode(
```

Comment:
I don't understand the name of this function at all, what Node is being provisioned? Having this all in one function makes complete sense to me though.

Reply:
The idea was that the node this code is running on would be provisioned.  I don't know if we have some better or more-consistent terminology for that, though.